androidenv: simplify updating, track latest, and recurseIntoAttrs all packages

numinit opened the pull request:
We can figure out the latest version for every package in mkrepo.rb, instead of repeating ourselves in compose-android-packages.

Track the latest non-preview versions of all packages by default, and remove version hardcoding everywhere. Enable the update script.

Only eval supported packages on aarch64-linux.

Depends on #374102 and #360151, and reduces the effort required to maintain androidenv packages by an order of magnitude.

CC @NixOS/android team!

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

numinit committed:
androidenv: simplify updating and track latest

We can figure out the latest version for every package in mkrepo.rb,
instead of repeating ourselves in compose-android-packages.

Track the latest non-preview versions of all packages by default, and
remove version hardcoding everywhere. Enable the update script.

numinit commented on pkgs/development/mobile/androidenv/cmdline-tools.nix:
```diff
@@ -46,5 +46,7 @@ deployAndroidPackage {
     ${postInstall}
   '';
 
+  passthru.updateScript = ./update.sh;
```
I still have to test with this a little more, hence this being a draft.

hadilq reviewed (approved):
It's awesome! Thank you!

numinit committed:
androidenv: mkrepo/fetchrepo: output update script commit object

numinit commented:
`NIXPKGS_ALLOW_UNFREE=1 NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE=1 nix-shell maintainers/scripts/update.nix --argstr package androidenv.test-suite --argstr commit true`

```
commit 3a419de475219862e46466a58791345743cbbb5d (HEAD -> androidenv/autoupdate)
Author: Morgan Jones <me@numin.it>
Date:   Thu Feb 20 08:13:18 2025 +0000

    androidenv: build-tools: 35.0.0 -> 35.0.1; cmdline-tools: 16.0 -> 17.0; cmake: 3.31.4 -> 3.31.5
    
    Performed the following automatic androidenv updates:
    
    - build-tools: 35.0.0 -> 35.0.1
    - cmdline-tools: 16.0 -> 17.0
    - cmake: 3.31.4 -> 3.31.5
    
    Tests exited with status: 0

diff --git a/pkgs/development/mobile/androidenv/repo.json b/pkgs/development/mobile/androidenv/repo.json
index 388aa1ac9a93..b6314fdb85a1 100644
--- a/pkgs/development/mobile/androidenv/repo.json
+++ b/pkgs/development/mobile/androidenv/repo.json
@@ -13900,9 +13900,9 @@
     }
   },
   "latest": {
-    "build-tools": "35.0.0",
-    "cmake": "3.31.4",
-    "cmdline-tools": "16.0",
+    "build-tools": "35.0.1",
+    "cmake": "3.31.5",
+    "cmdline-tools": "17.0",
     "emulator": "35.4.8",
     "ndk": "28.0.13004108",
     "ndk-bundle": "22.1.7171670",
```

numinit commented:
I finally got tired of all the formatting... but made sure it's a separate commit.

JohnRTitor commented:
Looks like `aarch64-linux` fails to eval.

numinit commented:
Yeah, that one is expected. In #360151 we added eval warnings if Google doesn't have packages for the platform, which aarch64-linux does not.

numinit committed:
androidenv: support if-supported for include* options

This lets us stop hardcoding platforms and start filtering for which
archives have sources, which is the most accurate way to do it.

Tested on x86_64-linux, aarch64-linux, and aarch64-darwin.

numinit commented on pkgs/development/mobile/androidenv/deploy-androidpackages.nix:
```diff
@@ -58,15 +58,20 @@ let
   pname = "android-sdk-${lib.concatMapStringsSep "-" (package: package.name) sortedPackages}";
   version = lib.concatMapStringsSep "-" (package: package.revision) sortedPackages;
-  src = map (package: package.archives) packages;
+  src = lib.flatten (map (package: package.archives) packages);
```
This was ending up like `src = [ [] ]`

numinit commented:
Fixed the eval warnings on aarch64-linux and other unsupported platforms by:

- pushing them into the package build step as errors
- omitting them from the Android composition if includeNDK/includeEmulator/etc are set to "if-supported" or null, so you can still have an Android composition with the SDKs if you want
- writing tests that are aware of which platforms things should work on

Checked on the aarch64 Linux and Darwin community builders, as well as my x86_64-linux.

Now we can do:

`NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE=1 NIXPKGS_ALLOW_UNFREE=1 nix-build -A androidenv`

numinit committed:
doc: update androidenv docs

numinit committed:
release-notes: add androidenv details to rl-2505

numinit commented:
Docs should be accurate now.

numinit commented:
`nix-build -A androidenv` is no longer a good idea, though it does technically work. Try it. :-)

numinit commented:
I did a nixfmt pass, though a lot of the attrsets and function args in this module could be cleaned up too. That'll be a separate PR.

numinit committed:
androidenv: add recurseIntoAttrs to packages

numinit commented:
CC @NixOS/android - I think I've done what I need to in order to make all the packages discoverable in the package search. Let me know if we should have a different approach.

numinit committed:
androidenv: nixfmt

hadilq commented on pkgs/development/mobile/androidenv/compose-android-packages.nix:
```diff
@@ -1,410 +1,707 @@
+    assert builtins.isList packages;
+    if check == true then
+      true
+    else if check == false then
```
Not sure but shouldn't `check == null` give use the same behavior as `check == false`?

numinit replied:
I'll probably just remove `null` as an option.

hadilq reviewed (approved):
Awesome work! Thank you!

JohnRTitor commented on pkgs/development/mobile/androidenv/tools.nix:
```diff
@@ -1,17 +1,48 @@
+      freetype
+      fontconfig
+      fontconfig.lib
+      stdenv.cc.cc.libgcc or null # fix for https://github.com/NixOS/nixpkgs/issues/226357
```
do we still need the fix?

hadilq replied:
The `tools` is an old package that is obsoleted from the Google side, so we probably have to remove it in near future. I would not touch it, especially that this change is only related to formatting.

JohnRTitor commented on pkgs/development/mobile/androidenv/mkrepo.rb:
Oh yes Ruby (my nightmare)

numinit committed:
androidenv: compose-android-packages: remove null synonym to if-supported

numinit committed:
doc: clarify how if-supported works

numinit commented:
Think everything should be addressed with the latest two commits.

numinit committed:
androidenv: accept license during tests

numinit committed:
androidenv: support minPlatformVersion and maxPlatformVersion

With the release of Android API 36, it was getting rather annoying to
need to keep these updated. Allow just specifying the min and max
instead.

numinit commented:
Upstream issue tracker links:

https://issuetracker.google.com/issues/403946427

https://issuetracker.google.com/issues/403946428

This should be good to merge, I think.

JohnRTitor commented:
Okay, let me take a look finally got around to review it.

JohnRTitor reviewed (approved):
LGTM, will merge after nixpkgs review

JohnRTitor commented:
Looks like we are getting some hash mismatches and source 404?


```

numinit commented:
Hrm. I think we may just need to regenerate the repo.json with the update script.

JohnRTitor commented:
I am just wondering if anything can be done about the hash mismatch, if upstream decides to modify it without creating a new tag/release, it will just error out, without regular maintenance.

numinit commented:
I wonder if we can pick a "safer" set of system images for the android-studio-full package (i.e. the default configuration of androidenv.androidSdk gives all but the ones that prove to be routinely problematic). I have seen this happen rarely on a couple packages.

numinit committed:
androidenv: prefer local build for fetching packages

numinit commented:
@JohnRTitor I see this in that nixpkgs-review you posted:

`error: writing to file: No space left on device`

`Error: No space left on device : '/Users/runner/runners/2.323.0/_diag/blocks/13cddd55-888a-496c-8745-378e12896358_04a6bb19-1e3a-5c2a-49d9-4f595c30705c.1'`

If you nixpkgs-review all of androidenv, it will download all historical Android images and packages so I am not surprised some are broken... I think this is a different issue than the android-studio-full hashchange on darwin.

numinit committed:
androidenv: emulator: support 35.5.x

numinit committed:
androidenv: platforms: 35 -> 36; sources: 35 -> 36; build-tools: 35.0.1 -> 36.0.0; cmdline-tools: 17.0 -> 19.0; skiaparser: 6 -> 7; cmake: 3.31.5 -> 3.31.6; emulator: 35.4.8 -> 35.5.8

Performed the following automatic androidenv updates:

- platforms: 35 -> 36
- sources: 35 -> 36
- build-tools: 35.0.1 -> 36.0.0
- cmdline-tools: 17.0 -> 19.0
- skiaparser: 6 -> 7
- cmake: 3.31.5 -> 3.31.6
- emulator: 35.4.8 -> 35.5.8

Tests exited with status: 0

Last 100 lines of output:
```
          PosixPath('/nix/store/xyff06pkhki3qy1ls77w10s0v79c9il0-reproducible-builds.sh/lib'),
          PosixPath('/nix/store/z7k98578dfzi6l3hsvbivzm7hfqlk0zc-set-source-date-epoch-to-latest.sh/lib'),
          PosixPath('/nix/store/pilsssjjdxvdphlg2h19p0bfx5q0jzkn-strip.sh/lib'),
          PosixPath('/nix/store/8v6k283dpbc0qkdq81nb6mrxrgcb10i1-gcc-wrapper-14-20241116/lib'),
          PosixPath('/nix/store/r8gajll0yb0c08hd734lyj3jmqav3p9a-binutils-wrapper-2.43.1/lib'),
          PosixPath('/nix/store/ycli2afmijaqi9vmv6rwz4jbcrbly3x7-unzip-6.0/lib'),
          PosixPath('/nix/store/43jxq9r017vpwkmph5x4mphk38r8v3fp-make-shell-wrapper-hook/lib'),
          PosixPath('/nix/store/5z8ff1wfq3f5yd74jbv3g5vzd27cq0nh-die-hook/lib'),
          PosixPath('/nix/store/yip39c32hz27ix425p14q6pw7w8i598g-auto-patchelf-hook/lib'),
          PosixPath('/nix/store/214vrsiv784x6marqh27ips2r4p6pq34-auto-patchelf-0-unstable-2024-08-14/lib'),
          PosixPath('/nix/store/b1y4f1xab6svz8in8azp7c8bl5r6sx2s-binutils-wrapper-2.43.1/lib'),
          PosixPath('/nix/store/nk427kbyf9l6fpfdxb8hmrq2nlkl8nfl-patchelf-0.15.0/lib'),
          PosixPath('/nix/store/qjb61crwqjr60mgd75z8cfx0cmjgd1i9-update-autotools-gnu-config-scripts-hook/lib'),
          PosixPath('/nix/store/jjhw2phnaip4kg0qjas3x3fsaifi8y0w-no-broken-symlinks.sh/lib'),
          PosixPath('/nix/store/h9lc1dpi14z7is86ffhl3ld569138595-audit-tmpdir.sh/lib'),
          PosixPath('/nix/store/m54bmrhj6fqz8nds5zcj97w9s9bckc9v-compress-man-pages.sh/lib'),
          PosixPath('/nix/store/wgrbkkaldkrlrni33ccvm3b6vbxzb656-make-symlinks-relative.sh/lib'),
          PosixPath('/nix/store/5yzw0vhkyszf2d179m0qfkgxmp5wjjx4-move-docs.sh/lib'),
          PosixPath('/nix/store/fyaryjvghbkpfnsyw97hb3lyb37s1pd6-move-lib64.sh/lib'),
          PosixPath('/nix/store/kd4xwxjpjxi71jkm6ka0np72if9rm3y0-move-sbin.sh/lib'),
          PosixPath('/nix/store/pag6l61paj1dc9sv15l7bm5c17xn5kyk-move-systemd-user-units.sh/lib'),
          PosixPath('/nix/store/cmzya9irvxzlkh7lfy6i82gbp0saxqj3-multiple-outputs.sh/lib'),
          PosixPath('/nix/store/hxv896faph0rqxjq2ycxpcrbnngc95sz-patch-shebangs.sh/lib'),
          PosixPath('/nix/store/cickvswrvann041nqxb0rxilc46svw1n-prune-libtool-files.sh/lib'),
          PosixPath('/nix/store/xyff06pkhki3qy1ls77w10s0v79c9il0-reproducible-builds.sh/lib'),
          PosixPath('/nix/store/z7k98578dfzi6l3hsvbivzm7hfqlk0zc-set-source-date-epoch-to-latest.sh/lib'),
          PosixPath('/nix/store/pilsssjjdxvdphlg2h19p0bfx5q0jzkn-strip.sh/lib'),
          PosixPath('/nix/store/8v6k283dpbc0qkdq81nb6mrxrgcb10i1-gcc-wrapper-14-20241116/lib'),
          PosixPath('/nix/store/r8gajll0yb0c08hd734lyj3jmqav3p9a-binutils-wrapper-2.43.1/lib')],
 'paths': [PosixPath('.')],
 'recursive': True,
 'runtime_dependencies': []}
auto-patchelf: 0 dependencies could not be satisfied
patching script interpreter paths in /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin
/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin/.screenshot2-wrapped: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/58br4vk3q5akf4g8lx0pqzfhn47k3j8d-bash-5.2p37/bin/sh"
/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin/.lint-wrapped: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/58br4vk3q5akf4g8lx0pqzfhn47k3j8d-bash-5.2p37/bin/sh"
/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin/.avdmanager-wrapped: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/58br4vk3q5akf4g8lx0pqzfhn47k3j8d-bash-5.2p37/bin/sh"
/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin/.d8-wrapped: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/58br4vk3q5akf4g8lx0pqzfhn47k3j8d-bash-5.2p37/bin/sh"
/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin/.sdkmanager-wrapped: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/58br4vk3q5akf4g8lx0pqzfhn47k3j8d-bash-5.2p37/bin/sh"
/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin/.r8-wrapped: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/58br4vk3q5akf4g8lx0pqzfhn47k3j8d-bash-5.2p37/bin/sh"
/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin/.profgen-wrapped: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/58br4vk3q5akf4g8lx0pqzfhn47k3j8d-bash-5.2p37/bin/sh"
/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin/.apkanalyzer-wrapped: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/58br4vk3q5akf4g8lx0pqzfhn47k3j8d-bash-5.2p37/bin/sh"
/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin/.resourceshrinker-wrapped: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/58br4vk3q5akf4g8lx0pqzfhn47k3j8d-bash-5.2p37/bin/sh"
/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/cmdline-tools/19.0/bin/.retrace-wrapped: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/58br4vk3q5akf4g8lx0pqzfhn47k3j8d-bash-5.2p37/bin/sh"
[=======================================] 100% Fetch remote repository...
Auto-selecting single ABI x86_64
Running phase: fixupPhase
checking for references to /build/ in /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk...
patching script interpreter paths in /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/apkanalyzer to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.retrace-wrapped to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.r8-wrapped to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.screenshot2-wrapped to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/retrace to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.d8-wrapped to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.resourceshrinker-wrapped to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.sdkmanager-wrapped_ to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.profgen-wrapped to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/d8 to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/lint to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.lint-wrapped to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/profgen to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.apkanalyzer-wrapped to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/avdmanager to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.avdmanager-wrapped to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/screenshot2 to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/.sdkmanager-wrapped to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/resourceshrinker to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/sdkmanager to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
rewriting symlink /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/bin/r8 to be relative to /nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk
building '/nix/store/q2gng1mpi5n05nd47rjndkyar7ix11py-shell-sdkmanager-licenses-test.drv'...
building '/nix/store/dgj4jknvdwj240hkilp3bjny3qhfryfm-shell-sdkmanager-packages-test.drv'...
Do you wish to create a custom hardware profile? [no]
AVD 'testAVD' deleted.
Warning: Observed package id 'ndk;28.0.13004108' in inconsistent location '/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/ndk-bundle' (Expected '/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/ndk/28.0.13004108')
Warning: Observed package id 'ndk;28.0.13004108' in inconsistent location '/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/ndk-bundle' (Expected '/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/ndk/28.0.13004108')
Warning: Observed package id 'ndk;28.0.13004108' in inconsistent location '/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/ndk-bundle' (Expected '/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/ndk/28.0.13004108')
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-6.xml
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-5.xml
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-4.xml
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-3.xml
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-2.xml
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-1.xml
Warning: Failed to download any source lists!
Warning: IO exception while downloading manifest
Warning: IO exception while downloading manifest
Warning: Observed package id 'ndk;28.0.13004108' in inconsistent location '/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/ndk-bundle' (Expected '/nix/store/pxy9rcwwpp1r54azmh9aqw3i8pin2m9k-androidsdk/libexec/android-sdk/ndk/28.0.13004108')
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-6.xml
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-5.xml
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-4.xml
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-3.xml
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-2.xml
Warning: Failed to connect to host: https://dl.google.com/android/repository/addons_list-1.xml
Warning: Failed to download any source lists!
Warning: IO exception while downloading manifest
Warning: IO exception while downloading manifest
Warning: Still waiting for package manifests to be fetched remotely.
Warning: Still waiting for package manifests to be fetched remotely.
building '/nix/store/0dhv47xy0dv9dhshkd1cakm815zvm5kd-androidenv-test-suite.drv'...
/nix/store/1498sa9cvh8fkp1gm7s6kc5ivrrjy099-androidenv-test-suite
```

JohnRTitor commented:
Same issue [in this run](https://github.com/JohnRTitor/nixpkgs-review-gha/actions/runs/14104317074/job/39507275729), was trying to build only `androidenv.test-suite` and `android-studio-full`, it failed halfway to `androidenv.test-suite` due to insufficient space. (turns out 128 GB is not enough! yeah I am not running that locally)

I will start another run with just `android-studio-full`, I won't find time to run this locally unfortunately.

JohnRTitor commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 379534`

Logs: https://github.com/JohnRTitor/nixpkgs-review-gha/actions/runs/14104317074


---
### `x86_64-linux`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>androidenv.test-suite</li>
  </ul>
</details>


---
### `aarch64-linux`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>androidenv.test-suite</li>
  </ul>
</details>


---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>androidenv.test-suite</li>
  </ul>
</details>


---
### `aarch64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>androidenv.test-suite</li>
  </ul>
</details>

JohnRTitor commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 379534`

Logs: https://github.com/JohnRTitor/nixpkgs-review-gha/actions/runs/14105633661


---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>android-studio-full</li>
  </ul>
</details>


---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>android-studio-full</li>
  </ul>
</details>


---
### `x86_64-darwin`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>android-studio-full</li>
  </ul>
</details>


---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>android-studio-full</li>
  </ul>
</details>

JohnRTitor commented:
Huh? is it because of the config requirement?

numinit commented:
Time for me to cut down the system images then :-)

numinit committed:
androidenv: support 5 years of Android APIs with images; test 10 without

Cut down on the number of system images that need to be fetched by just
supporting packages from the last 5 years. Still test the last 10 years
of APIs without images.

numinit commented:
Alright, that should be a little more reasonable. The example shells were quite exhaustive, and now that we support better control over which API levels get pulled in we can just say "last 5 API levels" which is roughly the last 5 years.

JohnRTitor commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 379534`

Logs: https://github.com/JohnRTitor/nixpkgs-review-gha/actions/runs/14144226401


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>androidenv.test-suite</li>
  </ul>
</details>


---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>androidenv.test-suite</li>
  </ul>
</details>


---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>androidenv.test-suite</li>
  </ul>
</details>


---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>androidenv.test-suite</li>
  </ul>
</details>

JohnRTitor commented:
![image](https://github.com/user-attachments/assets/c3549c77-d738-43f7-a14f-9bce6ef624ed)

![image](https://github.com/user-attachments/assets/210ed14f-3b6c-45c2-b463-18614cb87fe4)

numinit commented:
Nice, thanks for the review and merge. This one was really interesting. None of it was built in CI, and the parts that were about to be were giant and all manually tested. Glad to get all this figured out.

alapshin commented:
Just wanted to chime in and say big thank you for your work on android tooling in nixpkgs.

numinit commented:
@alapshin No problem, appreciate it! As I've been learning more nixpkgs things, especially codebases like this one, I've been trying to go back and make the practices used better when I discover a more reliable way. I still want to be able to replace things like the prepackaged NDK and cmake with things we build, that may be up next.

Atemu commented:
This broke orgzly's android env:

https://github.com/orgzly-revived/orgzly-android-revived/blob/dee9cfbdf824e6796257ff7566479fb16193a824/flake.nix#L23-L41

I narrowed it down to this happenening when more than one ndk version is declared:

```
ln: failed to create symbolic link 'ndk/28.0.13004108/ndk-bundle': Permission denied
```

28 is not one of the declared ones, so I assuem it's the version in nixpkgs

I'm not sure whether you're still supposed to declare all of what orgzly declares here; I don't know the androidenv code. If not, this should probably also be said in the release notes.

numinit commented:
Thanks, taking a look at this.

numinit commented:
@Atemu I tested your particular Android composition against #402391, that breaking was a regression and it is fixed with that PR.
